### PR TITLE
[keymgr] keymgr hardening part 3

### PR DESCRIPTION
**Pull request by tjaychen**
- Protect kmac transfer counter
- Sparse encoding of kmac transfer fsm.
- Make the fault / error behavior more consistent.  Faults will always cause keymgr to transition into "Disabled" state if an operation is ongoing.  When transitioning because of a fault, the key state is always "wiped" with entropy.  Once transitioned into disable state, the behavior will always be consistent, ie, random data will be used for operations, and the resulting garbage data is also used to update the state.

### CODE REVIEW

**Comment by imphil** [comment]
(Leaving this review to others. Please let me know if you explicitly want my input, @tjaychen!)

**Comment by tjaychen** [comment]
@imphil input is always appreciated!

File: hw/ip/keymgr/data/keymgr.hjson [hunk 1]
```diff
@@ -387,23 +387,63 @@
...
-        { bits: "0",
+        { bits: "2:0",
           name: "VAL",
           resval: "0"
           desc: '''
-            When 1, the sideload keys are continuously cleared with entropy.
-            When 0, the sideload keys are left as is.
-          '''
+            Depending on the value programmed, a different side load key slot is cleared.
+            If the value programmed is not one of the enumerated values below, ALL side load key slots are continuously cleard.
```
Comment (sriyerg) [inline hw/ip/keymgr/data/keymgr.hjson:399]:
```suggestion
            If the value programmed is not one of the enumerated values below, ALL side load key slots are continuously cleared.
```

Perhaps limit to 100 chars / line?

**Review by weicaiyang — APPROVED** [review]
LGTM

**Comment by tjaychen** [comment]
thanks for the review @weicaiyang ! I also noticed during this PR that the fault / error handling seemed a bit inconsistent and tried to clean that up a bit. I imagine if it's the way it is now, it will make DV somewhat painful.

When you do the DV for this part, we should also talk about how difficult it is to verify using a "timed" entropy reseed instead of an event based reseed (the timed one I feel makes it hard for DV to predict what the loaded entropy is).  If this becomes an issue, we can tweak the design to make it more friendly for DV.

**Comment by weicaiyang** [comment]
> thanks for the review @weicaiyang ! I also noticed during this PR that the fault / error handling seemed a bit inconsistent and tried to clean that up a bit. I imagine if it's the way it is now, it will make DV somewhat painful.
> 
> When you do the DV for this part, we should also talk about how difficult it is to verify using a "timed" entropy reseed instead of an event based reseed (the timed one I feel makes it hard for DV to predict what the loaded entropy is). If this becomes an issue, we can tweak the design to make it more friendly for DV.

@tjaychen I don't really check the entropy value or data generated from entropy. If entropy is used to wipe secret, I only compare wiped data don't match with any previous valid secret.
Hence, if we wipe data immediately when fault occurs or we transition to a Disable state, then it should be fine to me. Please correct me if this isn't the cases or I miss something.

**Comment by tjaychen** [comment]
that should be fine, but i wonder if we need to be more accurate in the DV modeling... ie specifically separate out the cases where it was wiped with entropy (thus not matching anything previously) and updated with garbage data (which can be controlled on the kmac interface).

What do you think?

**Comment by weicaiyang** [comment]
> that should be fine, but i wonder if we need to be more accurate in the DV modeling... ie specifically separate out the cases where it was wiped with entropy (thus not matching anything previously) and updated with garbage data (which can be controlled on the kmac interface).
> 
> What do you think?

it's very hard and painful to model garbage data since entropy interface is async, keymgr has a timer to do reseed and we also need to know the exact time keymgr use entropy. Seems not worth making this accurate model.

**Comment by tjaychen** [comment]
> > that should be fine, but i wonder if we need to be more accurate in the DV modeling... ie specifically separate out the cases where it was wiped with entropy (thus not matching anything previously) and updated with garbage data (which can be controlled on the kmac interface).
> > What do you think?
> 
> it's very hard and painful to model garbage data since entropy interface is async, keymgr has a timer to do reseed and we also need to know the exact time keymgr use entropy. Seems not worth making this accurate model.

yeah in general i agree. But i'm trying to think if there's any reason to explicitly verify whether something is wiped by entropy, or "processed" entropy.  I'll have to think on this.

**Comment by weicaiyang** [comment]
> > > that should be fine, but i wonder if we need to be more accurate in the DV modeling... ie specifically separate out the cases where it was wiped with entropy (thus not matching anything previously) and updated with garbage data (which can be controlled on the kmac interface).
> > > What do you think?
> > 
> > 
> > it's very hard and painful to model garbage data since entropy interface is async, keymgr has a timer to do reseed and we also need to know the exact time keymgr use entropy. Seems not worth making this accurate model.
> 
> yeah in general i agree. But i'm trying to think if there's any reason to explicitly verify whether something is wiped by entropy, or "processed" entropy. I'll have to think on this.

please LMK if you find something interesting to verify. I was thinking if our design just uses pseudo random value from LFSR without using data from entropy, we actually don't know and so far, no way to check that.

**Pull request merged**